### ISSUE TITLE
[PATCH _63] Slurping empty files

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#1002]
Migrated from [rt.perl.org#1952](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=1952) (status was 'resolved')

Searchable as RT1952$

**Comment by p5pRT** [comment]
### From hansmu@xs4all.nl

Slurping an empty file should produce a defined empty string\.

In current perls \(\_03 and \_63\)\, this only happens if $\. == 0\,
for example&#8203;:

&nbsp;    undef $/;                 
&nbsp;    open F\, "/etc/passwd";
&nbsp;    \<F>;    \# set $\. = 1
&nbsp;    open F\, "/dev/null";
&nbsp;    print defined\(\<F>\) ? "ok\\n" : "not ok\\n";

prints&#8203;:

&nbsp;    not ok

A comment in pp\_hot\.c suggests that the IOf\_NOLINE flag
should flip\-flop when it is used\.  It doesn't do that\.
For one thing\, $\. is incremented when a defined empty
string is returned and this implies the IOf\_NOLINE flag
isn't looked at again until the file handle is closed\.

The patch below clears the IOf\_NOLINE flag when a handle is
opened and sets it on each successful read\.  It removes the
flip\-flopping and the reference to $\.
The result is that you can slurp a defined empty string off
an empty file once\, irresepective of $\. \.

\-\- HansM

\-\-\- doio\.c\.orig    Wed Dec  8 07&#8203;:23&#8203;:10 1999
\*\*\* doio\.c    Wed Dec 29 02&#8203;:44&#8203;:07 1999
@&#8203;@&#8203; \-452\,6 \+452\,7 @&#8203;@&#8203;
&nbsp;     \}
 \#endif
&nbsp;     IoIFP\(io\) = fp;
\+    IoFLAGS\(io\) &= ~IOf\_NOLINE;
&nbsp;     if \(writing\) \{
&nbsp;     dTHR;
&nbsp;     if \(IoTYPE\(io\) == 's'
\-\-\- pp\_hot\.c\.orig    Wed Dec  8 07&#8203;:23&#8203;:14 1999
\*\*\* pp\_hot\.c    Wed Dec 29 03&#8203;:07&#8203;:35 1999
@&#8203;@&#8203; \-1284\,12 \+1284\,10 @&#8203;@&#8203;
&nbsp;     offset = 0;
&nbsp;     \}
 
\-/\* flip\-flop EOF state for a snarfed empty file \*/
\+/\* delay EOF state for a snarfed empty file \*/
 \#define SNARF\_EOF\(gimme\,rs\,io\,sv\) \\
\-    \(\(gimme \!= G\_SCALAR || SvCUR\(sv\)                    \\
\-      || \(IoFLAGS\(io\) & IOf\_NOLINE\) || IoLINES\(io\) || \!RsSNARF\(rs\)\)    \\
\-    ? \(\(IoFLAGS\(io\) &= ~IOf\_NOLINE\)\, TRUE\)                \\
\-    : \(\(IoFLAGS\(io\) |= IOf\_NOLINE\)\, FALSE\)\)
\+    \(gimme \!= G\_SCALAR || SvCUR\(sv\)            \\
\+      || \(IoFLAGS\(io\) & IOf\_NOLINE\) || \!RsSNARF\(rs\)\)
 
&nbsp;     for \(;;\) \{
&nbsp;     if \(\!sv\_gets\(sv\, fp\, offset\)
@&#8203;@&#8203; \-1322\,6 \+1320\,7 @&#8203;@&#8203;
&nbsp;         SvTAINTED\_on\(sv\);
&nbsp;     \}
&nbsp;     IoLINES\(io\)\+\+;
\+    IoFLAGS\(io\) |= IOf\_NOLINE;
&nbsp;     SvSETMAGIC\(sv\);
&nbsp;     XPUSHs\(sv\);
&nbsp;     if \(type == OP\_GLOB\) \{



---
Site configuration information for perl 5\.00563&#8203;:

Configured by hansm at Mon Dec 13 02&#8203;:03&#8203;:09 MET 1999\.

Summary of my perl5 \(revision 5\.0 version 5 subversion 63\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=next\, osvers=4\_2\, archname=OPENSTEP\-Mach
&nbsp;    uname='bombadil '
&nbsp;    config\_args='\-des \-Dcf\_email=hansmu@&#8203;xs4all\.nl \-Dprefix=/usr/local \-Doptimize=\-g \-O'
&nbsp;    hint=recommended\, useposix=undef\, d\_sigaction=undef
&nbsp;    usethreads=undef useperlio=undef d\_sfio=undef
&nbsp;    use64bits=undef usemultiplicity=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, optimize='\-g \-O'\, gccversion=NeXT DevKit\-based CPP 4\.0
&nbsp;    cppflags='\-dynamic \-fno\-common \-DUSE\_NEXT\_CTYPE \-DUSE\_PERL\_SBRK \-arch m68k \-DDEBUGGING \-I/usr/local/include'
&nbsp;    ccflags ='\-dynamic \-fno\-common \-DUSE\_NEXT\_CTYPE \-DUSE\_PERL\_SBRK \-arch m68k \-arch i386 \-DDEBUGGING \-I/usr/local/include'
&nbsp;    stdchar='char'\, d\_stdstdio=define\, usevfork=false
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    alignbytes=8\, usemymalloc=y\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-dynamic \-prebind  \-arch m68k \-arch i386 \-L/usr/local/lib'
&nbsp;    libpth=/lib /usr/lib /usr/local/lib
&nbsp;    libs= 
&nbsp;    libc=/NextLibrary/Frameworks/System\.framework/System\, so=dylib\, useshrplib=true\, libperl=libperl\.5\.dylib
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_next\.xs\, dlext=bundle\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\, lddlflags=' \-dynamic \-bundle \-undefined suppress  \-arch m68k \-arch i386 \-L/usr/local/lib'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl 5\.00563&#8203;:
&nbsp;    lib
&nbsp;    /Users/hansm/lib/perl
&nbsp;    /usr/local/lib/perl5/5\.00563/OPENSTEP\-Mach
&nbsp;    /usr/local/lib/perl5/5\.00563
&nbsp;    /usr/local/lib/site\_perl/5\.00563/OPENSTEP\-Mach
&nbsp;    /usr/local/lib/site\_perl
&nbsp;    \.



---
Environment for perl 5\.00563&#8203;:
&nbsp;    DYLD\_LIBRARY\_PATH=/Users/hansm/src/perl/build/perl\-5\.006/perl5\.005\_63
&nbsp;    HOME=/Users/hansm
&nbsp;    LANG \(unset\)
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/Users/hansm/bin&#8203;:/usr/local/bin&#8203;:/usr/games&#8203;:/usr/ucb&#8203;:/bin&#8203;:/usr/bin&#8203;:/usr/etc&#8203;:/Users/hansm/bin/cookies&#8203;:/LocalApps/Opener\.app&#8203;:\.
&nbsp;    PERL5LIB=/Users/hansm/lib/perl
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/usr/bin/zsh

**Comment by p5pRT** [comment]
### From @gsar

On Wed\, 29 Dec 1999 03&#8203;:38&#8203;:04 \+0100\, Hans Mulder wrote&#8203;:
>Slurping an empty file should produce a defined empty string\.
>
>In current perls \(\_03 and \_63\)\, this only happens if $\. == 0\,
>for example&#8203;:
>
>    undef $/;                 
>    open F\, "/etc/passwd";
>    \<F>;    \# set $\. = 1
>    open F\, "/dev/null";
>    print defined\(\<F>\) ? "ok\\n" : "not ok\\n";
>
>prints&#8203;:
>
>    not ok
>
>A comment in pp\_hot\.c suggests that the IOf\_NOLINE flag
>should flip\-flop when it is used\.  It doesn't do that\.
>For one thing\, $\. is incremented when a defined empty
>string is returned and this implies the IOf\_NOLINE flag
>isn't looked at again until the file handle is closed\.

The comment is certainly misleading\.  As I recall\, the IoLINES\(\)
check was added mostly as an afterthought to not make it a flip\-flop
as seen from Perl \(as described in perldelta\)\, but it appears I
overlooked the fact that IoLINES\(\) is not reset by an implicit
close\.  \(This means the example in perldelta won't work for more
than one file on the command line either\.\)

>The patch below clears the IOf\_NOLINE flag when a handle is
>opened and sets it on each successful read\.  It removes the
>flip\-flopping and the reference to $\.
>The result is that you can slurp a defined empty string off
>an empty file once\, irresepective of $\. \.

The reason the flag was set/cleared in SNARF\_EOF\(\) was to avoid
adding yet more code to the hot read loop\.  I'd like to preserve
that if possible\.

I think your patch effectively does&#8203;:

 \-      || \(IoFLAGS\(io\) & IOf\_NOLINE\) || IoLINES\(io\) || \!RsSNARF\(rs\)\)    \\
 \+      || \(IoFLAGS\(io\) & IOf\_NOLINE\) || \!RsSNARF\(rs\)\)    \\

apart from making it an once\-only thing rather than a flip\-flop\.
Any reason why we can't do just the above?

Thanks\.


Sarathy
gsar@&#8203;ActiveState\.com

**Comment by p5pRT** [comment]
### From @gsar

On Wed\, 29 Dec 1999 11&#8203;:28&#8203;:50 PST\, I wrote&#8203;:
>On Wed\, 29 Dec 1999 03&#8203;:38&#8203;:04 \+0100\, Hans Mulder wrote&#8203;:
>>A comment in pp\_hot\.c suggests that the IOf\_NOLINE flag
>>should flip\-flop when it is used\.  It doesn't do that\.
>>For one thing\, $\. is incremented when a defined empty
>>string is returned and this implies the IOf\_NOLINE flag
>>isn't looked at again until the file handle is closed\.
>
>The comment is certainly misleading\.  As I recall\, the IoLINES\(\)
>check was added mostly as an afterthought to not make it a flip\-flop
>as seen from Perl \(as described in perldelta\)\, but it appears I
>overlooked the fact that IoLINES\(\) is not reset by an implicit
>close\.  \(This means the example in perldelta won't work for more
>than one file on the command line either\.\)
>
>>The patch below clears the IOf\_NOLINE flag when a handle is
>>opened and sets it on each successful read\.  It removes the
>>flip\-flopping and the reference to $\.
>>The result is that you can slurp a defined empty string off
>>an empty file once\, irresepective of $\. \.
>
>The reason the flag was set/cleared in SNARF\_EOF\(\) was to avoid
>adding yet more code to the hot read loop\.  I'd like to preserve
>that if possible\.
>
>I think your patch effectively does&#8203;:
>
> \-      || \(IoFLAGS\(io\) & IOf\_NOLINE\) || IoLINES\(io\) || \!RsSNARF\(rs\)\)    \\
> \+      || \(IoFLAGS\(io\) & IOf\_NOLINE\) || \!RsSNARF\(rs\)\)    \\
>
>apart from making it an once\-only thing rather than a flip\-flop\.
>Any reason why we can't do just the above?

This here is what I'm talking about\.


Sarathy
gsar@&#8203;ActiveState\.com
<details><summary>Inline Patch</summary>

```diff
-----------------------------------8<-----------------------------------
Change 4736 by gsar@auger on 1999/12/29 21:04:59

	slurp mode fix in change#2910 wasn't quite right (spotted by Hans
	Mulder)

Affected files ...

... //depot/perl/doio.c#89 edit
... //depot/perl/pp_hot.c#151 edit
... //depot/perl/t/io/argv.t#10 edit

Differences ...

==== //depot/perl/doio.c#89 (text) ====
Index: perl/doio.c
--- perl/doio.c.~1~	Wed Dec 29 13:05:03 1999
+++ perl/doio.c	Wed Dec 29 13:05:03 1999
@@ -452,6 +452,7 @@
     }
 #endif
     IoIFP(io) = fp;
+    IoFLAGS(io) &= ~IOf_NOLINE;
     if (writing) {
 	dTHR;
 	if (IoTYPE(io) == 's'

==== //depot/perl/pp_hot.c#151 (text) ====
Index: perl/pp_hot.c
--- perl/pp_hot.c.~1~	Wed Dec 29 13:05:03 1999
+++ perl/pp_hot.c	Wed Dec 29 13:05:03 1999
@@ -1283,12 +1283,11 @@
 	offset = 0;
     }
 
-/* flip-flop EOF state for a snarfed empty file */
+/* delay EOF state for a snarfed empty file */
 #define SNARF_EOF(gimme,rs,io,sv) \
-    ((gimme != G_SCALAR || SvCUR(sv)					\
-      || (IoFLAGS(io) & IOf_NOLINE) || IoLINES(io) || !RsSNARF(rs))	\
-	? ((IoFLAGS(io) &= ~IOf_NOLINE), TRUE)				\
-	: ((IoFLAGS(io) |= IOf_NOLINE), FALSE))
+    (gimme != G_SCALAR || SvCUR(sv)					\
+     || !RsSNARF(rs) || (IoFLAGS(io) & IOf_NOLINE)			\
+     || ((IoFLAGS(io) |= IOf_NOLINE), FALSE))
 
     for (;;) {
 	if (!sv_gets(sv, fp, offset)

==== //depot/perl/t/io/argv.t#10 (xtext) ====
Index: perl/t/io/argv.t
--- perl/t/io/argv.t.~1~	Wed Dec 29 13:05:03 1999
+++ perl/t/io/argv.t	Wed Dec 29 13:05:03 1999
@@ -5,29 +5,29 @@
     unshift @INC, '../lib';
 }
 
-print "1..14\n";
+print "1..20\n";
 
 use File::Spec;
 
 my $devnull = File::Spec->devnull;
 
-open(try, '>Io.argv.tmp') || (die "Can't open temp file: $!");
+open(try, '>Io_argv1.tmp') || (die "Can't open temp file: $!");
 print try "a line\n";
 close try;
 
 if ($^O eq 'MSWin32') {
-  $x = `.\\perl -e "while (<>) {print \$.,\$_;}" Io.argv.tmp Io.argv.tmp`;
+  $x = `.\\perl -e "while (<>) {print \$.,\$_;}" Io_argv1.tmp Io_argv1.tmp`;
 }
 else {
-  $x = `./perl -e 'while (<>) {print \$.,\$_;}' Io.argv.tmp Io.argv.tmp`;
+  $x = `./perl -e 'while (<>) {print \$.,\$_;}' Io_argv1.tmp Io_argv1.tmp`;
 }
 if ($x eq "1a line\n2a line\n") {print "ok 1\n";} else {print "not ok 1\n";}
 
 if ($^O eq 'MSWin32') {
-  $x = `.\\perl -le "print 'foo'" | .\\perl -e "while (<>) {print \$_;}" Io.argv.tmp -`;
+  $x = `.\\perl -le "print 'foo'" | .\\perl -e "while (<>) {print \$_;}" Io_argv1.tmp -`;
 }
 else {
-  $x = `echo foo|./perl -e 'while (<>) {print $_;}' Io.argv.tmp -`;
+  $x = `echo foo|./perl -e 'while (<>) {print $_;}' Io_argv1.tmp -`;
 }
 if ($x eq "a line\nfoo\n") {print "ok 2\n";} else {print "not ok 2\n";}
 
@@ -39,7 +39,7 @@
 }
 if ($x eq "foo\n") {print "ok 3\n";} else {print "not ok 3 :$x:\n";}
 
-@ARGV = ('Io.argv.tmp', 'Io.argv.tmp', $devnull, 'Io.argv.tmp');
+@ARGV = ('Io_argv1.tmp', 'Io_argv1.tmp', $devnull, 'Io_argv1.tmp');
 while (<>) {
     $y .= $. . $_;
     if (eof()) {
@@ -52,49 +52,74 @@
 else
     {print "not ok 5\n";}
 
-open(try, '>Io.argv.tmp') or die "Can't open temp file: $!";
+open(try, '>Io_argv1.tmp') or die "Can't open temp file: $!";
+close try;
+open(try, '>Io_argv2.tmp') or die "Can't open temp file: $!";
 close try;
-@ARGV = 'Io.argv.tmp';
+@ARGV = ('Io_argv1.tmp', 'Io_argv2.tmp');
 $^I = '.bak';
 $/ = undef;
+my $i = 6;
 while (<>) {
-    s/^/ok 6\n/;
+    s/^/ok $i\n/;
+    ++$i;
     print;
 }
-open(try, '<Io.argv.tmp') or die "Can't open temp file: $!";
+open(try, '<Io_argv1.tmp') or die "Can't open temp file: $!";
+print while <try>;
+open(try, '<Io_argv2.tmp') or die "Can't open temp file: $!";
 print while <try>;
 close try;
 undef $^I;
 
 eof try or print 'not ';
-print "ok 7\n";
+print "ok 8\n";
 
 eof NEVEROPENED or print 'not ';
-print "ok 8\n";
+print "ok 9\n";
 
-open STDIN, 'Io.argv.tmp' or die $!;
+open STDIN, 'Io_argv1.tmp' or die $!;
 @ARGV = ();
 !eof() or print 'not ';
-print "ok 9\n";
+print "ok 10\n";
 
 <> eq "ok 6\n" or print 'not ';
-print "ok 10\n";
+print "ok 11\n";
 
 open STDIN, $devnull or die $!;
 @ARGV = ();
 eof() or print 'not ';
-print "ok 11\n";
+print "ok 12\n";
 
-@ARGV = ('Io.argv.tmp');
+@ARGV = ('Io_argv1.tmp');
 !eof() or print 'not ';
-print "ok 12\n";
+print "ok 13\n";
 
 @ARGV = ($devnull, $devnull);
 !eof() or print 'not ';
-print "ok 13\n";
+print "ok 14\n";
 
 close ARGV or die $!;
 eof() or print 'not ';
-print "ok 14\n";
+print "ok 15\n";
+
+{
+    local $/;
+    open F, 'Io_argv1.tmp' or die;
+    <F>;	# set $. = 1
+    open F, $devnull or die;
+    print "not " unless defined(<F>);
+    print "ok 16\n";
+    print "not " if defined(<F>);
+    print "ok 17\n";
+    print "not " if defined(<F>);
+    print "ok 18\n";
+    open F, $devnull or die;	# restart cycle again
+    print "not " unless defined(<F>);
+    print "ok 19\n";
+    print "not " if defined(<F>);
+    print "ok 20\n";
+    close F;
+}
 
-END { unlink 'Io.argv.tmp', 'Io.argv.tmp.bak' }
+END { unlink 'Io_argv1.tmp', 'Io_argv1.tmp.bak', 'Io_argv2.tmp', 'Io_argv2.tmp.bak' }
End of Patch.
```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Gurusamy Sarathy wrote&#8203;:

> I think your patch effectively does&#8203;:

> \-      || \(IoFLAGS\(io\) & IOf\_NOLINE\) || IoLINES\(io\) || \!RsSNARF\(rs\)\)    \\
> \+      || \(IoFLAGS\(io\) & IOf\_NOLINE\) || \!RsSNARF\(rs\)\)    \\

> Any reason why we can't do just the above?

That would change behaviour&#8203;: in \_03\, if you read a number
of lines with a defined $/ and then slurp the rest of the
file\, you get undef if the rest happens to be 0 bytes long\.
With your change\, you'd get a defined empty string\.

I think that this would be an improvement as it would be more
consistent with the behaviour for empty files\.  But we may have
to explain the issue to the Backward Compatibility Police\.

And I think we also want to do

\+    IoFLAGS\(io\) &= ~IOf\_NOLINE;

somewhere near the end of Perl\_do\_open9\(\)\.

\-\- HansM
